File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/CacheDialogFragment.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

import com.wa2c.android.medoly.plugin.action.lrclyrics.R;
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCache;
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCacheHelper;
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.ResultItem;

import java.lang.ref.WeakReference;
import java.util.Collections;


/**
 * Confirm dialog.
 */
public class CacheDialogFragment extends AbstractDialogFragment {

    /** Dialog result on delete lyrics. */
    public final static int DIALOG_RESULT_DELETE_LYRICS = -10;
    /** Dialog result on delete cache. */
    public final static int DIALOG_RESULT_DELETE_CACHE = -20;

    /** Cache key. */
    private final static String ARG_CACHE = "ARG_CACHE";

    /**
     * Create dialog instance.
     * @param cache Search cache.
     * @return Dialog instance.
     */
    static public CacheDialogFragment newInstance(SearchCache cache) {
        CacheDialogFragment fragment = new CacheDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_CACHE, cache);
        fragment.setArguments(args);

        return fragment;
    }



    /**
     * onCreateDialog
     */
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreateDialog(savedInstanceState);

        // data
        final SearchCache cache = (SearchCache)getArguments().getSerializable(ARG_CACHE);
        ResultItem result = cache.makeResultItem();

        // view
        final View content = View.inflate(getActivity(), R.layout.dialog_cache, null);
        TextView textView = (TextView)content.findViewById(R.id.dialogCacheLyricsTextView);
        textView.setText((cache.has_lyrics != null && cache.has_lyrics) ? result.getLyrics() : getString(R.string.message_dialog_cache_none));

        // delete lyrics button
        content.findViewById(R.id.dialogCacheDeleteLyricsButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new LyricsDeleteAsyncTask(CacheDialogFragment.this, cache).execute();
            }
        });

        // delete cache button
        content.findViewById(R.id.dialogCacheDeleteCacheButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new CacheDeleteAsyncTask(CacheDialogFragment.this, cache).execute();
            }
        });

        // build dialog
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(R.string.title_activity_cache);
        builder.setView(content);
        builder.setNeutralButton(R.string.label_close, null);
        builder.setNegativeButton(R.string.label_dialog_cache_research, clickListener);
        if (result != null && !TextUtils.isEmpty(result.getLyrics())) {
            builder.setPositiveButton(R.string.menu_search_save_file, clickListener);
        }
        return  builder.create();
    }

    /**
     * Delete lyrics task
     */
    private static class LyricsDeleteAsyncTask extends AsyncTask<Void, Void, Void> {

        private WeakReference<AbstractDialogFragment> dialogReference;
        private SearchCache cache;

        LyricsDeleteAsyncTask(@NonNull AbstractDialogFragment dialog, SearchCache cache) {
            this.dialogReference = new WeakReference<>(dialog);
            this.cache = cache;
        }

        @Override
        protected Void doInBackground(Void... params) {
            SearchCacheHelper searchCacheHelper = new SearchCacheHelper(dialogReference.get().getActivity());
            searchCacheHelper.insertOrUpdate(cache.title, cache.artist, null);
            return null;
        }
        @Override
        protected void onPostExecute(Void result) {
            dialogReference.get().onClickButton(DIALOG_RESULT_DELETE_LYRICS);
        }
    }

    /**
     * Delete lyrics task
     */
    private static class CacheDeleteAsyncTask extends AsyncTask<Void, Void, Void> {

        private WeakReference<AbstractDialogFragment> dialogReference;
        private SearchCache cache;

        CacheDeleteAsyncTask(@NonNull AbstractDialogFragment dialog, SearchCache cache) {
            this.dialogReference = new WeakReference<>(dialog);
            this.cache = cache;
        }

        @Override
        protected Void doInBackground(Void... params) {
            SearchCacheHelper searchCacheHelper = new SearchCacheHelper(dialogReference.get().getActivity());
            searchCacheHelper.delete(Collections.singletonList(cache));
            return null;
        }
        @Override
        protected void onPostExecute(Void result) {
            dialogReference.get().onClickButton(DIALOG_RESULT_DELETE_CACHE);
        }
    }
}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/CacheDialogFragment.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import com.wa2c.android.medoly.plugin.action.lrclyrics.R
import com.wa2c.android.medoly.plugin.action.lrclyrics.databinding.DialogCacheBinding
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.AppDatabase
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCache
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCacheDao
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch


/**
 * Cache dialog.
 */
class CacheDialogFragment : AbstractDialogFragment() {

    /** Binding. */
    private lateinit var binding: DialogCacheBinding
    private lateinit var dao: SearchCacheDao

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.dialog_cache, null, false)
        dao = AppDatabase.buildDb(context).getSearchCacheDao()

        // data
        val cache = arguments!!.getSerializable(ARG_CACHE) as SearchCache
        val result = cache.makeResultItem()

        // view
        binding.dialogCacheLyricsTextView.text = if (cache.has_lyrics != null && cache.has_lyrics!!) result!!.lyrics else getString(R.string.message_dialog_cache_none)
        binding.dialogCacheDeleteLyricsButton.isEnabled = (cache.has_lyrics == true)

        // delete lyrics button
        binding.dialogCacheDeleteLyricsButton.setOnClickListener {
            deleteLyrics(cache)
        }

        // delete cache button
        binding.dialogCacheDeleteCacheButton.setOnClickListener {
            deleteCache(cache)
        }

        // build dialog
        return AlertDialog.Builder(context).apply {
            setTitle(R.string.title_activity_cache)
            setView(binding.root)
            setNeutralButton(R.string.label_close, null)
            setNegativeButton(R.string.label_dialog_cache_research, null)
            result?.let {
                if (!result.lyrics.isNullOrEmpty())
                    setPositiveButton(R.string.menu_search_save_file, null)
            }
            create()
        }.create()
    }

    /**
     * Delete cache lyrics.
     */
    private fun deleteLyrics(cache: SearchCache) {
        GlobalScope.launch(Dispatchers.Main) {
            val deleteResult = async(Dispatchers.Default) {
                try {
                    cache.language = null
                    cache.from = null
                    cache.file_name = null
                    cache.has_lyrics = null
                    cache.result = null
                    cache.date_modified = System.currentTimeMillis()
                    dao.update(cache)
                    return@async true
                } catch (e: Exception) {
                    return@async null
                }
            }
            val r = deleteResult.await()
            if (r != true)
                AppUtils.showToast(context, R.string.message_dialog_cache_delete_error)
            invokeListener(DIALOG_RESULT_DELETE_LYRICS)
        }
    }

    /**
     * Delete cache.
     */
    private fun deleteCache(cache: SearchCache) {
        GlobalScope.launch(Dispatchers.Main) {
            val deleteResult = async(Dispatchers.Default) {
                try {
                    dao.delete(cache)
                    return@async true
                } catch (e: Exception) {
                    return@async null
                }
            }
            val r = deleteResult.await()
            if (r != true)
                AppUtils.showToast(context, R.string.message_dialog_cache_delete_error)
            invokeListener(DIALOG_RESULT_DELETE_CACHE)
        }
    }

    companion object {

        /** Dialog result on delete lyrics.  */
        const val DIALOG_RESULT_DELETE_LYRICS = -10
        /** Dialog result on delete cache.  */
        const val DIALOG_RESULT_DELETE_CACHE = -20

        /** Cache key.  */
        private const val ARG_CACHE = "ARG_CACHE"

        /**
         * Create dialog instance.
         * @param cache Search cache.
         * @return Dialog instance.
         */
        fun newInstance(cache: SearchCache): CacheDialogFragment {
            return CacheDialogFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(ARG_CACHE, cache)
                }
            }
        }
    }
}
